motion/thruster_interface_auv_cpp/src/thruster_interface_auv_driver.cpp

```diff
@@ -0,0 +1,138 @@
+    PWM_MIN(PWM_MIN),
+    PWM_MAX(PWM_MAX),
+    COEFFS(COEFFS)
+{
+
+    // Convert SYSTEM_OPERATIONAL_VOLTAGE passed as argument to assign the integer field variable [one is double, field var is int]
+    if (SYSTEM_OPERATIONAL_VOLTAGE < 11.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 10;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 13.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 12;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 15.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 14;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 17.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 16;
+    } else if (SYSTEM_OPERATIONAL_VOLTAGE < 19.0) {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 18;
+    } else {
+        this->SYSTEM_OPERATIONAL_VOLTAGE = 20;
+    }
+
+    // Open the I2C bus
+    std::string i2c_filename = "/dev/i2c-" + std::to_string(I2C_BUS);
+    bus_fd = open(i2c_filename.c_str(), O_RDWR);
+    if (bus_fd < 0) {
+        std::cerr << "ERROR: Failed to open I2C bus " << I2C_BUS << std::endl;
+    }
+}
+
+ThrusterInterfaceAUVDriver::~ThrusterInterfaceAUVDriver() {
+    if (bus_fd >= 0) {
+        close(bus_fd);
+    }
+}
+
+std::vector<int16_t> ThrusterInterfaceAUVDriver::interpolate_forces_to_pwm(const std::vector<double>& thruster_forces_array) {
+    // Convert Newtons to Kg (since the thruster datasheet is in Kg)
+    std::vector<double> forces_in_kg(thruster_forces_array.size());
+    for (size_t i = 0; i < thruster_forces_array.size(); ++i) {
+        forces_in_kg[i] = thruster_forces_array[i] / 9.80665;
+    }
+
+    // Select the appropriate coefficients based on the operational voltage
+    auto left_coeffs = COEFFS[SYSTEM_OPERATIONAL_VOLTAGE]["LEFT"];
+    auto right_coeffs = COEFFS[SYSTEM_OPERATIONAL_VOLTAGE]["RIGHT"];
+
+    std::vector<int16_t> interpolated_pwm;
+    for (size_t i = 0; i < forces_in_kg.size(); ++i) {
+        double force = forces_in_kg[i];
+        double pwm = 0.0;
+        if (force < 0) {
+            pwm =   left_coeffs[0] * std::pow(forces_in_kg[i], 3) + \
+                    left_coeffs[1] * std::pow(forces_in_kg[i], 2) + \
+                    left_coeffs[2] * forces_in_kg[i] + \
+                    left_coeffs[3];
+        } else if (force == 0.0) {
+            pwm = 1500;
+        } else {
+            pwm =   right_coeffs[0] * std::pow(forces_in_kg[i], 3) + \
+                    right_coeffs[1] * std::pow(forces_in_kg[i], 2) + \
+                    right_coeffs[2] * forces_in_kg[i] + \
+                    right_coeffs[3];
+        }
+        interpolated_pwm.push_back(static_cast<int16_t>(pwm));
+    }
+
+    return interpolated_pwm;
+}
+
+void ThrusterInterfaceAUVDriver::send_data_to_escs(const std::vector<int16_t>& thruster_pwm_array) {
+    uint8_t i2c_data_array[16];
+    for (size_t i = 0; i < thruster_pwm_array.size(); ++i) {
+        i2c_data_array[2*i] = (thruster_pwm_array[i] >> 8) & 0xFF; // MSB
+        i2c_data_array[2*i + 1] = thruster_pwm_array[i] & 0xFF;    // LSB
+    }
```

Comment:
Suggestion:

```C++
constexpr std::size_t i2c_data_size = 8 * 2;  // (8 thrusters * (MSB + LSB)
std::uint8_t i2c_data_array[i2c_data_size];
auto pwm_to_i2c_data = [](std::int16_t pwm) -> std::array<std::uint8_t, 2> {
    return {static_cast<std::uint8_t>((pwm >> 8) & 0xFF), static_cast<std::uint8_t>(pwm & 0xFF)};
};
std::transform(thruster_pwm_array.begin(), thruster_pwm_array.end(), i2c_data_array, pwm_to_i2c_data);
```

Reply:
Tried to follow the suggestion but dived in compilation errors that i dont know how to fix. Apparently the 8-bt pair has to be manually flattened otherwise the cast from 8-bit pair to 16bit singleton doesnt end fine and i'm having `error: cannot convert ‘std::array<unsigned char, 2>’ to ‘unsigned char’ in assignment`. Trying to use copilot or gpt suggests me longer verbose ways to fix the code that requires inserting more loop and/or complicates even more the std::transform call making the code not readable at all.

Reply:
update: solved using this:
```C++
 constexpr std::size_t i2c_data_size = 8 * 2;  // 8 thrusters * (1xMSB + 1xLSB)
    std::vector<std::uint8_t> i2c_data_array;
    i2c_data_array.reserve(i2c_data_size);

    auto pwm_to_i2c_data = [](std::int16_t pwm) -> std::array<std::uint8_t, 2> {
        return {static_cast<std::uint8_t>((pwm >> 8) & 0xFF), static_cast<std::uint8_t>(pwm & 0xFF)};
    };

    std::for_each(thruster_pwm_array.begin(), thruster_pwm_array.end(), [&](std::int16_t pwm) {
        std::array<std::uint8_t,2> bytes = pwm_to_i2c_data(pwm);
        std::copy(bytes.begin(), bytes.end(), std::back_inserter(i2c_data_array));
    });
```

is it okay?